engineering-plans/0002-ethereum-tracing-cache.md

```diff
@@ -0,0 +1,201 @@
+
+### Limitations
+The problem of reorg turns out to be a particularly tricky one for the cache, mostly due to ranges of blocks being requested rather than individual hashes. To sidestep this problem, only blocks that are older than the reorg threshold will be eligible for caching.
+
+Additionally, there are some subgraphs which may require traces from all or a substantial number of blocks and don't make effective use of filtering. In particular, subgraphs which specify a call handler without a contract address fall into this category. In order to prevent the cache from bloating, any use of Ethereum traces which does not filter on a contract address will bypass the cache.
+
+### EthereumTraceCache
+
+The implementation introduces the following trait, which is implemented primarily by `Store`.
+
+```rust
+use std::ops::RangeToInclusive;
+struct TracesInRange {
+    range: RangeToInclusive<u64>,
+    traces: Vec<Trace>,
+}
+
+pub trait EthereumTraceCache: Send + Sync + 'static {
+    /// Attempts to retrieve traces from the cache. Returns ranges which were retrieved.
+    /// The results may not cover the entire range of blocks. It is up to the caller to decide
+    /// what to do with ranges of blocks that are not cached.
+    fn traces_for_blocks(contract_address: Option<H160>, blocks: Range<u64>
+        ) -> Box<dyn Future<Output=Result<Vec<TracesInRange>, Error>>>;
+    fn add(contract_address: Option<H160> traces: Vec<TracesInRange>);
+}
+```
+
+#### Block schema
+
+Each cached block will exist as its own row in the database in an `eth_traces_cache` table.
+
+```rust
+eth_traces_cache(id) {
+  id -> Integer,
+  network -> Text,
+  block_number: Integer,
+  contract_address: Bytea,
+  traces -> Jsonb,
+}
+```
+
+A multi-column index will be added on network, block_number, and contract_address.
+
+Additionally, an `eth_traces_meta` table will be added to eventually enable cleaning out old data. It is substantially similar to the the existing `eth_call_meta` table.
+
+```rust
+eth_traces_meta(id) {
+  id -> (network, contract_address),
+  accessed_at -> Date,
+}
+```
+
+It can be noted that in the `eth_traces_cache` table, there is a very low cardinality for the value of the network row. It is inefficient for example to store the string `mainnet` millions of times and consider this value when querying. A data oriented approach would be to partition these tables on the value of the network. It is expected that hash partitioning available in Postgres 11 would be useful here, but the necessary dependencies won't be ready in time for this RFC. This may be revisited in the future.
+
+#### Valid Cache Range
+Because the absence of trace data for a block is a valid cache result, the database must maintain a data structure indicating which ranges of the cache are valid.
+
+This is the schema for that structure:
+```rust
+id -> Integer,
+network -> Text,
+start_block -> Integer,
+end_block -> Integer,
+contract_address -> Nullable<Bytea>,
+```
```

Comment:
I wonder if it wouldn't make sense to merge this into the `eth_traces_meta` table?

Reply:
I don't understand.

Reply:
I was just thinking that
```rust
eth_traces_meta(id) {
  network -> Text,
  contract_address -> Bytea,
  accessed_at -> Date,
  network -> Text,
  start_block -> Integer,
  end_block -> Integer,
}
```
could work as well. The start/end blocks are also meta data in a way. Does that make more sense?

Reply:
I suppose that would work. I don't have any strong opinions on this part either way.

Reply:
Since there's already so many pieces to keep track of (not in this RFC but overall), I'd be in favor of using two instead of three tables, one for the traces themselves and one for the meta data and range information.

Reply:
Ok. I'll make that adjustment.